src/3d/mesh/qgsmesh3dgeometry_p.cpp

```diff
@@ -0,0 +1,421 @@
+                 abs( otherFunctor->mVertScale - mVertScale ) < std::numeric_limits<float>::min() );
+      return false;
+    }
+
+    QT3D_FUNCTOR( MeshPlaneVertexBufferFunctor )
+
+  private:
+
+    QgsTriangularMesh mMesh;
+    QgsRectangle mExtent;
+    float mVertScale;
+
+};
+
+//! Generates index buffer for Mesh
+class MeshPlaneIndexBufferFunctor : public QBufferDataGenerator
+{
+  public:
+    explicit MeshPlaneIndexBufferFunctor( const QgsTriangularMesh &mesh )
+      : mMesh( mesh )
+    {}
+
+    QByteArray operator()() final
+    {
+      return createPlaneIndexData( mMesh );
+    }
+
+    bool operator ==( const QBufferDataGenerator &other ) const final
+    {
+      const MeshPlaneIndexBufferFunctor *otherFunctor = functor_cast<MeshPlaneIndexBufferFunctor>( &other );
+      if ( otherFunctor != nullptr )
+        return ( otherFunctor->mMesh.triangles().count() == mMesh.triangles().count() );
+      return false;
+    }
+
+    QT3D_FUNCTOR( MeshPlaneIndexBufferFunctor )
+
+  private:
+    QgsTriangularMesh mMesh;
+};
+
+//! Generates vertex buffer for Mesh using dataset value as verticale magnitude and for color
+class MeshPlaneVertexDatasetBufferFunctor : public QBufferDataGenerator
+{
+  public:
+    explicit MeshPlaneVertexDatasetBufferFunctor( const QgsTriangularMesh &mesh,
+        const QVector<double> verticaleMagnitude,
+        const QVector<double> scalarMagnitude,
+        const QgsRectangle &extent,
+        float vertScale )
+      : mMesh( mesh ),
+        mVerticaleMagnitude( verticaleMagnitude ),
+        mScalarMagnitude( scalarMagnitude ),
+        mExtent( extent ),
+        mVertScale( vertScale )
+    {}
+
+    QByteArray operator()() final
+    {
+      return createPlaneVertexDatasetData( mMesh,
+                                           mVerticaleMagnitude,
+                                           mScalarMagnitude,
+                                           mExtent,
+                                           mVertScale );
+    }
+
+    bool operator ==( const QBufferDataGenerator &other ) const final
+    {
+      const MeshPlaneVertexDatasetBufferFunctor *otherFunctor = functor_cast<MeshPlaneVertexDatasetBufferFunctor>( &other );
+      if ( otherFunctor != nullptr )
+        return ( otherFunctor->mMesh.triangles().count() == mMesh.triangles().count() &&
+                 otherFunctor->mVerticaleMagnitude.count() == mVerticaleMagnitude.count() &&
+                 otherFunctor->mScalarMagnitude.count() == mScalarMagnitude.count() &&
+                 otherFunctor->mExtent == mExtent &&
+                 abs( otherFunctor->mVertScale - mVertScale ) < std::numeric_limits<float>::min() );
+      return false;
+    }
+
+    QT3D_FUNCTOR( MeshPlaneVertexBufferFunctor )
+
+  private:
+    QgsTriangularMesh mMesh;
+    QVector<double> mVerticaleMagnitude;
+    QVector<double> mScalarMagnitude;
+    QgsRectangle mExtent;
+    float mVertScale;
+
+};
+
+QgsMesh3dGeometry_p::QgsMesh3dGeometry_p(
```

Comment:
Can we remove the _p suffix from the class name? We only use the _p suffix for file names of files that are part of the private implementations, but not for class names...